File: src/test/resources/de/hhu/stups/codegenerator/range_big/BSet.cpp
#include <iostream>
#include <string>
#include <queue>
#include <vector>
#include <cstdarg>
#include <immer/set.hpp>
#include "BInteger.cpp"
#include "BString.cpp"
#include "BStruct.cpp"

#ifndef BSET_H
#define BSET_H

using namespace std;

template<typename T>
class BSet : public BObject {

    public:

        typedef BSet<T> current_type;
        typedef T value_type;
        typedef void left_type;
        typedef void right_type;

        struct Hash {
            public:
                size_t operator()(const T& obj) const {
                    return obj.hashCode();
                }
        };

        struct HashEqual {
            public:
                bool operator()(const T& obj1, const T& obj2) const {
                    if (obj1 == obj2)
                        return true;
                    else
                        return false;
                    }
            };

        /*Only used within this class*/
        BSet<T>(const immer::set<T, Hash, HashEqual>& elements) {
            this->set = elements;
        }

        template<typename... Args>
        BSet<T>(const Args&... args) {
          this->set = var(args...);
        }

        immer::set<T, Hash, HashEqual> var() {
          return immer::set<T,Hash, HashEqual>();
        }

        template<typename R, typename... Args>
        immer::set<R, Hash, HashEqual> var(const R& first, const Args&... args) {
          return var(args...).insert(first);
        }

        BSet<T>() {
            this->set = immer::set<T,Hash, HashEqual>();
        }

        /*BSet<T>(const BSet<T>& set) {
            this->set = set.set;
        }*/

	/*public BSet(java.util.Set<BObject> elements) {
		this.set = HashTreePSet.from(elements);
	}

	public BSet(PSet<BObject> elements) {
		this.set = elements;
	}

	public BSet(BObject... elements) {
		this.set = HashTreePSet.from(Arrays.asList(elements));
	}*/

	/*public static LinkedHashSet<BObject> newStorage() {
		return new LinkedHashSet<>();
	}

	public java.lang.String toString() {
		Iterator<BObject> it = this.iterator();
		StringBuffer sb = new StringBuffer();
		sb.append("{");
		while (it.hasNext()) {
			BObject b = (BObject) it.next();
			sb.append(b.toString());
			if (it.hasNext()) {
				sb.append(", ");
			}
		}
		sb.append("}");
		return sb.toString();
	}*/

        int size() const {
            return set.size();
        }

        bool isEmpty() {
            return set.empty();
        }

        bool contains(const T& o) const {
            return set.count(o) > 0;
        }

        /*boolean add(BObject bObject) {
            throw new UnsupportedOperationException();
        }

        boolean remove(Object o) {
            throw new UnsupportedOperationException();
        }

        void clear() {
            throw new UnsupportedOperationException();
        }*/

        /*boolean equals(Object o) {
            if (this == o)
                return true;
            if (o == null || getClass() != o.getClass())
                return false;

            BSet bObjects = (BSet) o;

            if (!set.equals(bObjects.set))
                return false;

            return true;
        }

        int hashCode() {
            return set.hashCode();
        }

        boolean removeAll(Collection<?> c) {
            throw new UnsupportedOperationException();
        }

        Object[] toArray() {
            return set.toArray();
        }

        <T> T[] toArray(T[] a) {
            return set.toArray(a);
        }

        boolean containsAll(Collection<?> c) {
            return set.containsAll(c);
        }

        boolean addAll(Collection<? extends BObject> c) {
            throw new UnsupportedOperationException();
        }

        boolean retainAll(Collection<?> c) {
            throw new UnsupportedOperationException();
        }

        Iterator<BObject> iterator() {
            return set.iterator();
        }*/


        BSet<T> intersect(const BSet<T>& set) const {
            immer::set<T,Hash, HashEqual> result = this->set;
            for (const T& obj : this->set) {
                if(set.set.count(obj) == 0) {
                    result = result.erase(obj);
                }
            }
            return BSet(result);
        }

     	template<typename K = value_type>
     	K intersect() const {
     	    if(this->size() == 0) {
     	        return K();
     	    }
     	    K result;
     	    int i = 0;
            for(const T& s : this->set) {
                if(i == 0) {
                    result = K(s);
                }
                ++i;
                result = result.intersect(K(s));
            }
            return result;
     	}

        BSet<T> difference(const BSet<T>& set) const {
            if(this->size() == 0) {
                return BSet(this->set);
            }
            immer::set<T, Hash, HashEqual> result = this->set;
            for (const T& obj : set.set) {
                if(result.count(obj) == 1) {
                    result = result.erase(obj);
                }
            }
            return BSet(result);
        }

        BSet<T> _union(const BSet<T>& set) const {
            int thisSize = this->size();
            int otherSize = set.size();
            if(thisSize > otherSize) {
                immer::set<T,Hash, HashEqual> result = this->set;
                for (const T& obj : set.set) {
                    if(result.count(obj) == 0) {
                        result = result.insert(obj);
                    }
                }
                return BSet(result);
            } else {
                immer::set<T,Hash, HashEqual> result = set.set;
                for (const T& obj : this->set) {
                    if(result.count(obj) == 0) {
                        result = result.insert(obj);
                    }
                }
                return BSet(result);
            }
        }

     	template<typename K = value_type>
     	K _union() const {
     	    K result;
            for(const T& s : this->set) {
                result = result._union(K(s));
            }
            return result;
     	}

        static BSet<BInteger> interval(const BInteger& a, const BInteger& b) {
            immer::set<BInteger, Hash, HashEqual> result;
            int end = b.intValue();
            for(int i = a.intValue(); i <= end; ++i) {
                result = result.insert(i);
            }
            return BSet(result);
        }


        BInteger card() const {
            return BInteger(set.size());
        }

        BInteger _size() const {
            return BInteger(set.size());
        }

        BBoolean elementOf(const T& object) const {
            return BBoolean(set.count(object) > 0);
        }

        BBoolean notElementOf(const T& object) const {
            return BBoolean(set.count(object) == 0);
        }

        BBoolean subset(const BSet<T>& set) const {
            for (const T& obj : this->set) {
                if(set.set.count(obj) == 0) {
                    return BBoolean(false);
                }
            }
            return BBoolean(true);
        }

        BBoolean notSubset(const BSet<T>& set) const {
            for (const T& obj : this->set) {
                if(set.set.count(obj) == 0) {
                    return BBoolean(true);
                }
            }
            return BBoolean(false);
        }

        BBoolean strictSubset(const BSet<T>& set) const {
            if(this->set.size() != set.size()) {
                for (const T& obj : this->set) {
                    if(set.set.count(obj) == 0) {
                        return BBoolean(false);
                    }
                }
                return BBoolean(true);
            }
            return BBoolean(false);
        }

        BBoolean strictNotSubset(const BSet<T>& set) const {
            if(this->set.size() != set.size()) {
                for (const T& obj : this->set) {
                    if(set.set.count(obj) == 0) {
                        return BBoolean(true);
                    }
                }
                return BBoolean(false);
            }
            return BBoolean(false);
        }

        template<typename K = current_type>
    	BSet<K> pow() const {
    		BSet<K> result = BSet<K>();
    		K start = K();
    		queue<K> q = queue<K>();
    		q.push(start);
    		result = result._union(BSet<K>(start));
    		while(!q.empty()) {
    			K currentSet = q.front();
    			q.pop();
    			for(const T& element : this->set) {
    				K nextSet = currentSet._union(K(element));
    				int previousSize = result.size();
    				result = result._union(BSet<K>(nextSet));
    				if(previousSize < result.size()) {
    					q.push(nextSet);
    				}
    			}
    		}
    		return result;
    	}

        template<typename K = current_type>
    	BSet<K> pow1() const {
            K emptySet = K();
    		return this->pow().difference(BSet<K>(K()));
    	}

        template<typename K = current_type>
    	BSet<K> fin() const {
    		return this->pow();
    	}

        template<typename K = current_type>
    	BSet<K> fin1() const {
    		return this->pow1();
    	}

        BInteger min() const {
            BInteger result;
            int i = 0;
            for(const T& obj : this->set) {
                BInteger e = obj;
                if(i == 0) {
                    result = e;
                }
                if(e.less(result).booleanValue()) {
                    result = e;
                }
                ++i;
            }
            return result;
        }

        BInteger max() const {
            BInteger result;
            int i = 0;
            for(const T& obj : this->set) {
                BInteger e = obj;
                if(i == 0) {
                    result = e;
                }
                if(e.greater(result).booleanValue()) {
                    result = e;
                }
                ++i;
            }
            return result;
        }

        T nondeterminism() const {
		    int index = rand() % set.size();
		    typename immer::set<T,Hash, HashEqual>::const_iterator it = std::next(set.begin(), index);
		    return *it;
	    }

        friend bool operator !=(const BSet<T>& o1, const BSet<T>& o2) {
            return o1.set != o2.set;
        }

        friend bool operator ==(const BSet<T>& o1, const BSet<T>& o2) {
            return o1.set == o2.set;
        }

        BBoolean equal(const BSet<T>& other) const {
            return BBoolean(this->set == other.set);
        }

        BBoolean unequal(const BSet<T>& other) const {
            return BBoolean(this->set != other.set);
        }

        BBoolean subsetOfInteger() {
            for(T e : this->set) {
                if(typeid(e) == typeid(BInteger)) {
                    return BBoolean(true);
                } else {
                    return BBoolean(false);
                }
            }
            return BBoolean(true);
        }

        BBoolean strictSubsetOfInteger() {
            return this->subsetOfInteger();
        }

        BBoolean notSubsetOfInteger() {
            return this->subsetOfInteger()._not();
        }

        BBoolean equalInteger() {
            return BBoolean(false);
        }

        BBoolean unequalInteger() {
            return BBoolean(true);
        }

        BBoolean equalNatural() {
            return BBoolean(false);
        }

        BBoolean unequalNatural() {
            return BBoolean(true);
        }

        BBoolean equalNatural1() {
            return BBoolean(false);
        }

        BBoolean unequalNatural1() {
            return BBoolean(true);
        }

        BBoolean equalString() {
            return BBoolean(false);
        }

        BBoolean unequalString() {
            return BBoolean(true);
        }

        BBoolean equalStruct() {
            return BBoolean(false);
        }

        BBoolean unequalStruct() {
            return BBoolean(true);
        }

        BBoolean notStrictSubsetOfInteger() {
            return this->strictSubsetOfInteger()._not();
        }

        BBoolean subsetOfNatural() {
            for(T e : this->set) {
                BInteger element = (BInteger) e;
                if(!element.isNatural().booleanValue()) {
                    return BBoolean(false);
                }
            }
            return BBoolean(true);
        }

        BBoolean strictSubsetOfNatural() {
            return this->subsetOfNatural();
        }

        BBoolean notSubsetOfNatural() {
            return this->subsetOfNatural()._not();
        }

        BBoolean notStrictSubsetOfNatural() {
            return strictSubsetOfNatural()._not();
        }

        BBoolean subsetOfNatural1() {
            for(T e : this->set) {
                BInteger element = (BInteger) e;
                if(!element.isNatural1().booleanValue()) {
                    return BBoolean(false);
                }
            }
            return BBoolean(true);
        }

        BBoolean strictSubsetOfNatural1() {
            return this->subsetOfNatural1();
        }

        BBoolean notSubsetOfNatural1() {
            return this->subsetOfNatural1()._not();
        }

        BBoolean notStrictSubsetOfNatural1() {
            return this->strictSubsetOfNatural1()._not();
        }

        BBoolean subsetOfString() {
            for(T e : this->set) {
                BString element = (BString) e;
                if(!element.isString().booleanValue()) {
                    return BBoolean(false);
                }
            }
            return BBoolean(true);
        }

        BBoolean strictSubsetOfString() {
            return this->subsetOfString();
        }

        BBoolean notSubsetOfString() {
            return this->subsetOfString()._not();
        }

        BBoolean notStrictSubsetOfString() {
            return this->strictSubsetOfString()._not();
        }

        BBoolean subsetOfStruct() {
            for(T e : this->set) {
                BStruct element = (BStruct) e;
                if(!element.isRecord().booleanValue()) {
                    return BBoolean(false);
                }
            }
            return BBoolean(true);
        }

        BBoolean strictSubsetOfStruct() {
            return this->subsetOfStruct();
        }

        BBoolean notSubsetOfStruct() {
            return this->subsetOfStruct()._not();
        }

        BBoolean notStrictSubsetOfStruct() {
            return this->strictSubsetOfStruct()._not();
        }

        void operator =(const BSet<T>& other) {
            this->set = other.set;
        }

        int hashCode() const {
            int result = 0;
            int i = 0;
            for(const T& s : this->set) {
                if(i == 0) {
                    result = s.hashCode();
                }
                result = result ^ (s.hashCode() << 1);
                ++i;
            }
            return result;
        }

        friend std::ostream& operator<<(std::ostream &strm, const BSet<T>& set) {
            strm << "{";
            typename immer::set<T,Hash, HashEqual>::const_iterator it = set.begin();
            while(it != set.end()) {
                T element = *it;
                strm << element;
                ++it;
                if(it != set.end()) {
                    strm << ",";
                }
            }
            strm << "}";
            return strm;
        }

        typename immer::set<T,Hash, HashEqual>::const_iterator begin() const {
            return set.begin();
        }

        typename immer::set<T,Hash, HashEqual>::const_iterator end() const {
            return set.end();
        }

        immer::set<T,Hash, HashEqual> getSet() const {
            return set;
        }

        protected:
            immer::set<T,Hash, HashEqual> set;

};

#endif

File: src/test/resources/btypes/BSet.js
import { BBoolean } from "./BBoolean.js";
import { BInteger } from "./BInteger.js";
import { BRelation } from "./BRelation.js";
import { BString } from "./BString.js";
import { BStruct } from "./BStruct.js";
import * as immutable from "../immutable/dist/immutable.es.js";
export class BSet {
    constructor(...args) {
        if (args.length == 1 && args[0] instanceof immutable.Set) {
            this.set = args[0];
        }
        else {
            this.set = immutable.Set();
            for (let x of args) {
                this.set = this.set.add(x);
            }
        }
    }
    /* Make this class iterable */
    [Symbol.iterator]() {
        return this.set[Symbol.iterator]();
    }
    toString() {
        let sb = "{";
        this.set.forEach((element) => {
            if (sb.length > 1) {
                sb += ", ";
            }
            sb += element.toString();
        });
        sb += "}";
        return sb;
    }
    union(other) {
        let result = BSet.immutableSetUnion(this.set, other.set);
        return new BSet(result);
    }
    unionForSets() {
        if (this.set.isEmpty()) {
            return new BSet();
        }
        else {
            return this.set.reduce((a, e) => a.union(e), new BSet());
        }
    }
    unionForRelations() {
        if (this.set.isEmpty()) {
            return new BRelation();
        }
        else {
            return this.set.reduce((a, e) => a.union(e), new BRelation());
        }
    }
    static immutableSetUnion(s1, s2) {
        return s1.union(s2);
    }
    static immutableSetDifference(s1, s2) {
        return s1.subtract(s2);
    }
    static immutableSetIntersection(s1, s2) {
        return s1.intersect(s2);
    }
    static immutableMapUnion(m1, m2) {
        let result = m2;
        outer_loop: for (let current_element of m1.keys()) {
            for (let result_element of result.keys()) {
                if (current_element.equals(result_element)) {
                    let result_set = result.get(result_element);
                    let current_set = m1.get(current_element);
                    result = result.set(result_element, BSet.immutableSetUnion(result_set != null ? result_set : immutable.Set(), current_set != null ? current_set : immutable.Set()));
                    continue outer_loop;
                }
            }
            result = result.set(current_element, m1.get(current_element));
        }
        return result;
    }
    static immutableMapIntersection(m1, m2) {
        let result = immutable.Map();
        elem_loop: for (let result_element of m2.keys()) {
            for (let current_element of m1.keys()) {
                if (current_element.equals(result_element)) {
                    let result_set = m2.get(result_element);
                    let current_set = m1.get(current_element);
                    result = result.set(result_element, BSet.immutableSetIntersection(result_set != null ? result_set : immutable.Set(), current_set != null ? current_set : immutable.Set()));
                    continue elem_loop;
                }
            }
            result.delete(result_element);
        }
        return result;
    }
    intersect(other) {
        let result = BSet.immutableSetIntersection(this.set, other.set);
        return new BSet(result);
    }
    intersectForSets() {
        if (this.set.isEmpty()) {
            return new BSet();
        }
        else {
            return this.set.reduce((a, e) => a.intersect(e));
        }
    }
    intersectForRelations() {
        if (this.set.isEmpty()) {
            return new BRelation();
        }
        else {
            return this.set.reduce((a, e) => a.intersect(e));
        }
    }
    difference(other) {
        let set = BSet.immutableSetDifference(this.set, other.set);
        return new BSet(set);
    }
    card() {
        return new BInteger(this.set.size);
    }
    size() {
        return new BInteger(this.set.size);
    }
    elementOf(obj) {
        return new BBoolean(this.set.has(obj));
    }
    notElementOf(obj) {
        return new BBoolean(!this.set.has(obj));
    }
    subset(other) {
        for (let element of this.set) {
            if (!other.getSet().has(element)) {
                return new BBoolean(false);
            }
        }
        return new BBoolean(true);
    }
    notSubset(other) {
        return this.subset(other).not();
    }
    strictSubset(other) {
        return new BBoolean(other.getSet().size != this.set.size && this.subset(other).booleanValue());
    }
    strictNotSubset(other) {
        return this.strictSubset(other).not();
    }
    contains(other) {
        return other in this.set;
    }
    containsAll(other) {
        for (let o of other.set) {
            if (!(o in this.set)) {
                return false;
            }
        }
        return true;
    }
    isEmpty() {
        return this.set.size === 0;
    }
    equals(other) {
        if (!(other instanceof BSet)) {
            return false;
        }
        return this.set.equals(other.set);
    }
    equal(other) {
        return this.subset(other).and(other.subset(this));
    }
    unequal(other) {
        return this.equal(other).not();
    }
    nondeterminism() {
        let values = [];
        for (let item of this.set) {
            values.push(item);
        }
        return values[Math.floor(Math.random() * values.length)];
    }
    min() {
        return this.set.reduce((a, v) => { if (a < v) {
            return a;
        } return v; });
    }
    max() {
        return this.set.reduce((a, v) => { if (a > v) {
            return a;
        } return v; });
    }
    pow() {
        let result = new BSet();
        let start = new BSet();
        let queue = [start];
        result = result.union(new BSet(start));
        while (queue.length != 0) {
            let currentSet = queue.pop();
            for (let element of this.set) {
                let nextSet = currentSet.union(new BSet(element));
                let previousSize = result.size();
                result = result.union(new BSet(nextSet));
                if (previousSize < result.size()) {
                    queue.push(nextSet);
                }
            }
        }
        return result;
    }
    pow1() {
        return this.pow().difference(new BSet(new BSet()));
    }
    // Only finite subsets are supported so fin = pow
    fin() {
        return this.pow();
    }
    fin1() {
        return this.pow1();
    }
    subsetOfInteger() {
        for (let element of this.set) {
            if (element instanceof BInteger) {
                return new BBoolean(false);
            }
        }
        return new BBoolean(true);
    }
    strictSubsetOfInteger() {
        return this.subsetOfInteger();
    }
    notSubsetOfInteger() {
        return this.subsetOfInteger().not();
    }
    notStrictSubsetOfInteger() {
        return this.strictSubsetOfInteger().not();
    }
    subsetOfNatural() {
        for (let element of this.set) {
            if (!(element instanceof BInteger && element.isNatural().booleanValue())) {
                return new BBoolean(false);
            }
        }
        return new BBoolean(true);
    }
    strictSubsetOfNatural() {
        return this.subsetOfNatural();
    }
    notSubsetOfNatural() {
        return this.subsetOfNatural().not();
    }
    notStrictSubsetOfNatural() {
        return this.strictSubsetOfNatural().not();
    }
    subsetOfNatural1() {
        for (let element of this.set) {
            if (!(element instanceof BInteger && element.isNatural1().booleanValue())) {
                return new BBoolean(false);
            }
        }
        return new BBoolean(true);
    }
    subsetOfString() {
        for (let element of this.set) {
            if (!(element instanceof BString)) {
                return new BBoolean(false);
            }
        }
        return new BBoolean(true);
    }
    strictSubsetOfString() {
        return this.subsetOfString();
    }
    notSubsetOfString() {
        return this.subsetOfString().not();
    }
    notStrictSubsetOfString() {
        return this.strictSubsetOfString().not();
    }
    subsetOfStruct() {
        for (let element of this.set) {
            if (!(element instanceof BStruct)) {
                return new BBoolean(false);
            }
        }
        return new BBoolean(true);
    }
    strictSubsetOfStruct() {
        return this.subsetOfStruct();
    }
    notSubsetOfStruct() {
        return this.subsetOfStruct().not();
    }
    notStrictSubsetOfStruct() {
        return this.strictSubsetOfStruct().not();
    }
    equalInteger() {
        return new BBoolean(false);
    }
    unequalInteger() {
        return new BBoolean(true);
    }
    equalNatural() {
        return new BBoolean(false);
    }
    unequalNatural() {
        return new BBoolean(true);
    }
    equalNatural1() {
        return new BBoolean(false);
    }
    unequalNatural1() {
        return new BBoolean(true);
    }
    equalString() {
        return new BBoolean(false);
    }
    unequalString() {
        return new BBoolean(true);
    }
    equalStruct() {
        return new BBoolean(false);
    }
    unequalStruct() {
        return new BBoolean(true);
    }
    getSet() {
        return this.set;
    }
    static interval(a, b) {
        if (b.less(a).booleanValue()) {
            return new BSet();
        }
        let persistentSet = immutable.Set();
        for (let i = a; i.lessEqual(b).booleanValue(); i = i.plus(new BInteger(1))) {
            persistentSet = persistentSet.add(i);
        }
        return new BSet(persistentSet);
    }
    hashCode() {
        return this.set.hashCode();
    }
}
